x-pack/plugin/migrate/src/main/java/org/elasticsearch/xpack/migrate/ReindexDataStreamPipeline.java

```diff
@@ -0,0 +1,113 @@
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+package org.elasticsearch.xpack.migrate;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ingest.PutPipelineRequest;
+import org.elasticsearch.action.ingest.PutPipelineTransportAction;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.action.support.master.MasterNodeRequest;
+import org.elasticsearch.client.internal.Client;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.ingest.IngestMetadata;
+import org.elasticsearch.ingest.PipelineConfiguration;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentType;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+
+/**
+ * Manages the definitions and lifecycle of the ingest pipeline used by the reindex data stream operation.
+ */
+public class ReindexDataStreamPipeline {
+
+    /**
+     * The last version of the distribution that updated the pipeline definition.
+     */
+    static final int LAST_UPDATED_VERSION = Version.V_8_18_0.id;
+
+    public static final String PIPELINE_NAME = "reindex-data-stream";
+
+    /**
+     * Checks if the current version of the pipeline definition is installed in the cluster
+     * @param clusterState The cluster state to check
+     * @return true if a pipeline exists that is compatible with this version, false otherwise
+     */
+    public static boolean exists(ClusterState clusterState) {
+        final IngestMetadata ingestMetadata = clusterState.getMetadata().custom(IngestMetadata.TYPE);
+        // we ensure that we both have the pipeline and its version represents the current (or later) version
+        if (ingestMetadata != null) {
+            final PipelineConfiguration pipeline = ingestMetadata.getPipelines().get(PIPELINE_NAME);
+            if (pipeline != null) {
+                Object version = pipeline.getConfig().get("version");
+                // do not replace if pipeline was created by user and has no version
+                if (version == null) {
+                    return true;
+                }
+                return version instanceof Number number && number.intValue() >= LAST_UPDATED_VERSION;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Creates a pipeline with the current version's pipeline definition
+     * @param client Client used to execute put pipeline
+     * @param listener Callback used after pipeline has been created
+     */
+    public static void create(Client client, ActionListener<AcknowledgedResponse> listener) {
+        final BytesReference pipeline = BytesReference.bytes(currentPipelineDefinition());
+        client.execute(
+            PutPipelineTransportAction.TYPE,
+            new PutPipelineRequest(
```

Comment:
We probably ought to set the parent task id here so that it's cancellable (although I'm not 100% sure it's worth it).